src/ir/term/mod.rs

```diff
@@ -721,6 +723,104 @@ impl Array {
         self.check_idx(idx);
         self.map.get(idx).unwrap_or(&*self.default).clone()
     }
+
+    /// Iter
+    pub fn into_iter(&self) -> std::collections::btree_map::IntoIter<Value, Value> {
+        self.map.clone().into_iter()
+    }
+}
+
+/// Merge two Array Iterators
+pub struct ArrayMerge {
+    left: Peekable<Box<std::collections::btree_map::IntoIter<Value, Value>>>,
+    right: Peekable<Box<std::collections::btree_map::IntoIter<Value, Value>>>,
+    left_dfl: Value,
+    right_dfl: Value,
+}
+
+impl ArrayMerge {
+    /// Create a new [ArrayMerge] from two [Array]s
+    pub fn new(a: Array, b: Array) -> Self {
+        if a.size != b.size {
+            panic!("IR Arrays have different lengths: {}, {}", a.size, b.size);
+        }
+        if a.key_sort != b.key_sort {
+            panic!(
+                "IR Arrays have different key sorts: {}, {}",
+                a.key_sort, b.key_sort
+            );
+        }
+        if a.default.sort() != b.default.sort() {
+            panic!(
+                "IR Arrays default values have different key sorts: {}, {}",
+                a.default.sort(),
+                b.default.sort()
+            );
+        }
+
+        Self {
+            left: Box::new(a.into_iter()).peekable(),
+            right: Box::new(b.into_iter()).peekable(),
+            left_dfl: *a.default,
+            right_dfl: *b.default,
+        }
+    }
+
+    /// Iter
+    pub fn into_iter(&mut self) -> Box<dyn Iterator<Item = (Value, Value, Value)>> {
+        let mut acc: Vec<(Value, Value, Value)> = Vec::new();
+        let mut next = self.next_();
+        while let Some(n) = next {
+            acc.push(n);
+            next = self.next_();
+        }
+        Box::new(acc.into_iter())
+    }
+
+    /// Next
+    pub fn next_(&mut self) -> Option<(Value, Value, Value)> {
+        let l_peek = self.left.peek();
+        let r_peek = self.right.peek();
+
+        let mut left_next = false;
+        let mut right_next = false;
+
+        let res = match (l_peek, r_peek) {
+            (Some((l_ind, l_val)), Some((r_ind, r_val))) => match l_ind.cmp(r_ind) {
+                Ordering::Less => {
+                    left_next = true;
+                    Some((l_ind.clone(), l_val.clone(), self.right_dfl.clone()))
```

Comment:
More unnecessary cloning here, only to throw away `self.l.next()` below.

The right way to do this is to call `self.l.next()` right here and *use that value*. That also lets you get rid of the mutable state variables `left_next` and `right_next`, as I mentioned above.

```suggestion
                    let (l_ind, l_val) = self.l.next().unwrap(); // unwrap is safe because we peeked
                    Some((l_ind, l_val, self.right_dfl.clone()))
```

Of course, after fixing things `self.right_dfl` won't be cloned, it'll just be a reference.

Reply:
Of course, same comments apply to the remaining match clauses.